### PR TITLE
Fixed possible race when looking/registering an account

### PR DESCRIPTION
**Pull request by kozlovic**
Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

### CODE REVIEW

File: server/events_test.go [hunk 1]
```diff
@@ -1209,3 +1210,89 @@ func TestGatewayNameClientInfo(t *testing.T) {
...
+
+	nca, err := nats.Connect(urlA, createUserCreds(t, sa, akp))
+	if err != nil {
+		t.Fatalf("Error connecting to A: %v", err)
+	}
+	defer nca.Close()
+
+	// Since there is optimistic send, this message will go to B
+	// and on processing this message, B will lookup/fetch this
+	// account.
```
Comment (derekcollison) [inline server/events_test.go]: Should we say also that the system account will send a notification about the account to B as well?

File: server/events_test.go [hunk 2]
```diff
@@ -1209,3 +1210,89 @@ func TestGatewayNameClientInfo(t *testing.T) {
...
+	urlA := fmt.Sprintf("nats://%s:%d", oa.Host, oa.Port)
+	urlB := fmt.Sprintf("nats://%s:%d", ob.Host, ob.Port)
+
+	nca, err := nats.Connect(urlA, createUserCreds(t, sa, akp))
+	if err != nil {
+		t.Fatalf("Error connecting to A: %v", err)
+	}
+	defer nca.Close()
+
+	// Since there is optimistic send, this message will go to B
```
Comment (derekcollison) [inline server/events_test.go]: there is *an* optimistic

File: server/server.go [hunk 3]
```diff
@@ -820,7 +820,14 @@ func (s *Server) verifyAccountClaims(claimJWT string) (*jwt.AccountClaims, strin
 // Lock should be held upon entry.
 func (s *Server) fetchAccount(name string) (*Account, error) {
 	if accClaims, claimJWT, err := s.fetchAccountClaims(name); accClaims != nil {
-		acc := s.buildInternalAccount(accClaims)
+		// We have released the lock during the low level fetch.
+		// Now that we are back under lock, check again if account
+		// is in the map or not. If it is, simply return it.
+		acc := s.accounts[name]
```
Comment (derekcollison) [inline server/server.go]: Can be one line.

File: server/server.go [hunk 4]
```diff
@@ -820,7 +820,14 @@ func (s *Server) verifyAccountClaims(claimJWT string) (*jwt.AccountClaims, strin
...
+		if acc != nil {
+			return acc, nil
```
Comment (derekcollison) [inline server/server.go]: What about the case where we might have a newer claim here and may want to apply it?

File: server/events_test.go (hunk 1 above)
Comment (kozlovic) [reply server/events_test.go]: Updated

File: server/events_test.go (hunk 2 above)
Comment (kozlovic) [reply server/events_test.go]: Fixed

File: server/server.go (hunk 3 above)
Comment (kozlovic) [reply server/server.go]: done

File: server/server.go (hunk 4 above)
Comment (kozlovic) [reply server/server.go]: Not sure how to do that.. feel free to update PR or tell me what to do here.

Comment (derekcollison) [reply server/server.go]: ok will take a look. Working on another PR atm.

**Pull request closed without merging**